core/src/main/scala/kafka/common/ZkNodeChangeNotificationListener.scala

```diff
@@ -38,77 +39,80 @@ trait NotificationHandler {
  * notificationHandler's processNotification() method with the child's data as argument. As part of processing these changes it also
  * purges any children with currentTime - createTime > changeExpirationMs.
  *
- * The caller/user of this class should ensure that they use zkUtils.subscribeStateChanges and call processAllNotifications
- * method of this class from ZkStateChangeListener's handleNewSession() method. This is necessary to ensure that if zk session
- * is terminated and reestablished any missed notification will be processed immediately.
- * @param zkUtils
+ * @param zkClient
  * @param seqNodeRoot
- * @param seqNodePrefix
  * @param notificationHandler
  * @param changeExpirationMs
  * @param time
  */
-class ZkNodeChangeNotificationListener(private val zkUtils: ZkUtils,
+abstract class ZkNodeChangeNotificationListener(private val zkClient: KafkaZkClient,
```

Comment:
It seems that your early approach of using zkClient to get data from ZK directly in this class is simpler than making this an abstract class. Onur's comment is really not to expose AsyncResponse outside of KafkaZkClient.